File: packages/react-native/ios/Recording/AnrSentinel.swift
/**
 * Copyright 2026 Rejourney
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

@objc(AnrSentinel)
public final class AnrSentinel: NSObject {
    
    @objc public static let shared = AnrSentinel()
    
    private let _freezeThreshold: TimeInterval = 5.0
    private let _pollFrequency: TimeInterval = 2.0
    
    private var _watchThread: Thread?
    private var _volatile = VolatileState()
    private let _stateLock = os_unfair_lock_t.allocate(capacity: 1)
    
    private override init() {
        _stateLock.initialize(to: os_unfair_lock())
        super.init()
    }
    
    deinit {
        _stateLock.deallocate()
    }
    
    @objc public func activate() {
        os_unfair_lock_lock(_stateLock)
        guard _watchThread == nil else {
            os_unfair_lock_unlock(_stateLock)
            return
        }
        
        _volatile.running = true
        _volatile.lastResponse = Date().timeIntervalSince1970
        
        let t = Thread { [weak self] in self?._watchLoop() }
        t.name = "co.rejourney.anr"
        t.qualityOfService = .utility
        _watchThread = t
        os_unfair_lock_unlock(_stateLock)
        
        t.start()
    }
    
    @objc public func halt() {
        os_unfair_lock_lock(_stateLock)
        _volatile.running = false
        _watchThread = nil
        os_unfair_lock_unlock(_stateLock)
    }
    
    private func _watchLoop() {
        while true {
            os_unfair_lock_lock(_stateLock)
            let running = _volatile.running
            os_unfair_lock_unlock(_stateLock)
            guard running else { break }
            
            _sendPing()
            Thread.sleep(forTimeInterval: _pollFrequency)
            _checkPong()
        }
    }
    
    private func _sendPing() {
        os_unfair_lock_lock(_stateLock)
        if _volatile.awaitingPong {
            os_unfair_lock_unlock(_stateLock)
            return
        }
        _volatile.awaitingPong = true
        os_unfair_lock_unlock(_stateLock)
        
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            os_unfair_lock_lock(self._stateLock)
            self._volatile.lastResponse = Date().timeIntervalSince1970
            self._volatile.awaitingPong = false
            os_unfair_lock_unlock(self._stateLock)
        }
    }
    
    private func _checkPong() {
        os_unfair_lock_lock(_stateLock)
        let awaiting = _volatile.awaitingPong
        let last = _volatile.lastResponse
        os_unfair_lock_unlock(_stateLock)
        
        guard awaiting else { return }
        
        let delta = Date().timeIntervalSince1970 - last
        if delta >= _freezeThreshold {
            _reportFreeze(duration: delta)
        }
    }
    
    private func _reportFreeze(duration: TimeInterval) {
        DiagnosticLog.emit(.caution, "Main thread frozen for \(String(format: "%.1f", duration))s")
        
        ReplayOrchestrator.shared.incrementStalledTally()
        
        let trace = Thread.callStackSymbols.joined(separator: "\n")
        let ms = Int(duration * 1000)
        
        TelemetryPipeline.shared.recordAnrEvent(durationMs: ms, stack: trace)
    }
}

private struct VolatileState {
    var running = false
    var awaitingPong = false
    var lastResponse: TimeInterval = 0
}

@objc(ResponsivenessWatcher)
public final class ResponsivenessWatcher: NSObject {
    @objc public static let shared = ResponsivenessWatcher()
    
    private override init() { super.init() }
    
    @objc public func activate() {
        AnrSentinel.shared.activate()
    }
    
    @objc public func halt() {
        AnrSentinel.shared.halt()
    }
}

File: packages/react-native/android/src/main/java/com/rejourney/recording/AnrSentinel.kt
/**
 * Copyright 2026 Rejourney
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rejourney.recording

import android.os.Handler
import android.os.Looper
import com.rejourney.engine.DiagnosticLog
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong
import kotlin.concurrent.thread

/**
 * ANR (Application Not Responding) detection sentinel
 * Android implementation aligned with iOS AnrSentinel.swift
 * 
 * Uses a watchdog thread to detect main thread hangs > threshold
 */
class AnrSentinel private constructor() {
    
    companion object {
        @Volatile
        private var instance: AnrSentinel? = null
        
        val shared: AnrSentinel
            get() = instance ?: synchronized(this) {
                instance ?: AnrSentinel().also { instance = it }
            }
    }
    
    var currentSessionId: String? = null
    var anrThresholdMs: Long = 5000L
    
    private var watchdogThread: Thread? = null
    private val isActive = AtomicBoolean(false)
    private val lastResponseTime = AtomicLong(System.currentTimeMillis())
    private val pingSequence = AtomicInteger(0)
    private val pongSequence = AtomicInteger(0)
    
    private val mainHandler = Handler(Looper.getMainLooper())
    
    fun activate() {
        if (isActive.getAndSet(true)) return

        // Reset watchdog state on each activation to avoid stale timings from
        // previous app background periods.
        lastResponseTime.set(System.currentTimeMillis())
        pongSequence.set(pingSequence.get())
        
        startWatchdog()
    }
    
    fun deactivate() {
        if (!isActive.getAndSet(false)) return
        
        watchdogThread?.interrupt()
        watchdogThread = null
    }
    
    private fun startWatchdog() {
        watchdogThread = thread(name = "RJ-ANR-Watchdog", isDaemon = true) {
            val checkInterval = 1000L // 1 second
            
            while (isActive.get() && !Thread.currentThread().isInterrupted) {
                try {
                    // Send ping to main thread
                    val currentPing = pingSequence.incrementAndGet()
                    
                    mainHandler.post {
                        // Main thread is responsive, update pong
                        pongSequence.set(currentPing)
                        lastResponseTime.set(System.currentTimeMillis())
                    }
                    
                    Thread.sleep(checkInterval)
                    
                    // Check if main thread responded
                    val elapsed = System.currentTimeMillis() - lastResponseTime.get()
                    val missedPongs = pingSequence.get() - pongSequence.get()
                    
                    if (elapsed >= anrThresholdMs && missedPongs > 0) {
                        captureAnr(elapsed)
                        
                        // Reset to avoid duplicate reports
                        lastResponseTime.set(System.currentTimeMillis())
                        pongSequence.set(pingSequence.get())
                    }
                } catch (e: InterruptedException) {
                    Thread.currentThread().interrupt()
                    break
                } catch (e: Exception) {
                    DiagnosticLog.fault("ANR watchdog error: ${e.message}")
                }
            }
        }
    }
    
    private fun captureAnr(durationMs: Long) {
        try {
            val mainThread = Looper.getMainLooper().thread
            val stackTrace = mainThread.stackTrace
            
            val frames = stackTrace.map { element ->
                "${element.className}.${element.methodName}(${element.fileName}:${element.lineNumber})"
            }
            
            ReplayOrchestrator.shared?.incrementStalledTally()
            
            // Route ANR through TelemetryPipeline so it arrives in the events
            // batch and the backend ingest worker can insert it into the anrs table
            val stackStr = frames.joinToString("\n")
            TelemetryPipeline.shared?.recordAnrEvent(durationMs, stackStr)

            // Persist ANR incident and send through /api/ingest/fault so ANRs survive
            // process termination/background upload loss, similar to crash recovery.
            val sessionId = StabilityMonitor.shared?.currentSessionId
                ?: ReplayOrchestrator.shared?.replayId
                ?: "unknown"
            val incident = IncidentRecord(
                sessionId = sessionId,
                timestampMs = System.currentTimeMillis(),
                category = "anr",
                identifier = "MainThreadFrozen",
                detail = "Main thread unresponsive for ${durationMs}ms",
                frames = frames,
                context = mapOf(
                    "durationMs" to durationMs.toString(),
                    "threadState" to "blocked"
                )
            )
            StabilityMonitor.shared?.persistIncidentSync(incident)
            StabilityMonitor.shared?.transmitStoredReport()
            
            DiagnosticLog.fault("ANR detected: ${durationMs}ms hang")
            
        } catch (e: Exception) {
            DiagnosticLog.fault("Failed to capture ANR: ${e.message}")
        }
    }
}
